src/MacroTools/Frames/Books/ArtifactSystem/ArtifactBook.cs

```diff
@@ -51,12 +49,31 @@ private void OnArtifactDisposed(object? sender, Artifact artifact)
     {
       _pagesByArtifact[artifact].RemoveArtifact(artifact);
       _pagesByArtifact.Remove(artifact);
+      FillGaps();
+    }
+
+    private void FillGaps()
+    {
+        foreach (var artifact in _pagesByArtifact.Keys)
+        {
+            artifact.Disposed -= OnArtifactDisposed;
+        }
+        Pages.Clear();
```

Comment:
```cs
  foreach (var page in Pages) 
    page.Dispose();
```
You're clearing the frames out of memory but you're not disposing of them, so the pages don't get deleted until the garbage collector arrives. As far as I can tell this fixes the entire problem.

Reply:
Yes I figured that out too.
I am now disposing of the page and artifact frames properly I think.